### PR TITLE
pass instanceMetadata to updateNodeAddress to reduce api calls

### LINKED ISSUE #90652: Reduce the number of calls to GCE API (for clusters on GCE) coming from cloud node-controller
For external cloud provider - Kubelet setting the TaintExternalCloudProvider taint: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L266

The results in calling `initializeNode` function in cloud node-controller:
https://github.com/kubernetes/kubernetes/blob/5c276a451d64d161b21d70042d93f4c4b8e3201d/pkg/controller/cloud/node_controller.go#L323

This (in a healthy case) sends 4 GCE API calls to:
1. get node addresses: https://github.com/kubernetes/kubernetes/blob/5c276a451d64d161b21d70042d93f4c4b8e3201d/pkg/controller/cloud/node_controller.go#L439 (as part of getNodeModifier call)
2. get instance type:
https://github.com/kubernetes/kubernetes/blob/5c276a451d64d161b21d70042d93f4c4b8e3201d/pkg/controller/cloud/node_controller.go#L452 (as part of getNodeModifiers call)
3. get instance:
https://github.com/kubernetes/kubernetes/blob/5c276a451d64d161b21d70042d93f4c4b8e3201d/pkg/controller/cloud/node_controller.go#L233 (as part of updateNodeAddresses call)
4. get node addresses (again):
https://github.com/kubernetes/kubernetes/blob/5c276a451d64d161b21d70042d93f4c4b8e3201d/pkg/controller/cloud/node_controller.go#L242 (as part of updateNodeAddresses call)

All those requests are read requests and could be replaced by a single call.
We could do that by introducing a simple struct like:
```
type InstanceMetadata struct {
  ProviderID ...
  Type ...
  NodeAddresses ...
}
```
and adding a method (or rather replacing the existing InstanceTypeByProviderID which is used only for this purpose) to Instances cloud-provider interface;
```
func InstanceMetadataByProviderID(...) (*InstanceMetadata, error)
```

With that change, we would be able to replace all that 4 calls with just a single API call.

This would be useful especially for creating large clusters (like 5k nodes), which may saturate the quota for API calls for quite some time.

This is quite significant size of refactoring, though it should be relatively mechanical task to do.

/sig cloud-provider
/sig scalability
/priority important-longterm

/help-wanted

@mborsz @mm4tt - FYI

### PR DESCRIPTION
**Pull request by gongguan**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Pass instanceMetadata to updateNodeAddress to reduce api calls.

**Which issue(s) this PR fixes**:
Ref #90652

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

### CODE REVIEW

**Comment by wojtek-t** [comment]
LGTM, but unfortunately we are in code-freeze now and this needs to wait until we start accepting 1.20 PRs.

I would also like @andrewsykim to take a look.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go [hunk 1]
```diff
@@ -661,7 +661,7 @@ func (g *Cloud) Instances() (cloudprovider.Instances, bool) {
 
 // InstancesV2 returns an implementation of InstancesV2 for Google Compute Engine.
 func (g *Cloud) InstancesV2() (cloudprovider.InstancesV2, bool) {
-	return nil, false
+	return g, true
```
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: This should be done in a separate PR

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 2]
```diff
@@ -405,7 +405,7 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
 // a node object with provider-specific information.
 // All of the returned functions are idempotent, because they are used in a retry-if-conflict
 // loop, meaning they could get called multiple times.
-func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Context, node *v1.Node) ([]nodeModifier, error) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:408]: I don't think it makes sense for this method to be returning `InstanceMetadata`, what was the reason for returning it here?

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:408]: Returning `InstanceMetadata` and delivering it to `updateNodeAddress`, then `updateNodeAddress` won't call cloud apis to get `InstanceMetadata` again.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L393

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go (hunk 1 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: Since this change is tiny, I decided to put it in a separate commit now(this is also easy to revert).
How about my decision?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 2 above)
Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:408]: Could we get `InstanceMetadata` in a separate method before and pass it into `getNodeModifiersFromCloudProvider` instead?

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go (hunk 1 above)
Comment (andrewsykim) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: Prefer separate PR so it's more searchable and to keep the discussions specific to GCE in a separate PR thread

Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: ok, I'll do it.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 2 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:408]: Sure we can, but we should move get providerID outside `getNodeModifiersFromCloudProvider` at the same time.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L415-L436

Is it ok?

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:408]: I think so

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:408]: Done, PTAL.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go (hunk 1 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go]: Reverted this change.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 3]
```diff
@@ -405,63 +417,17 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
...
-	var (
-		nodeModifiers []nodeModifier
-		providerID    string
-		err           error
-	)
+func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(
+	ctx context.Context,
+	node *v1.Node,
+	providerID string,
+	providerIDModifier nodeModifier,
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:424]: remove this and just pass in getProviderID directly below?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 4]
```diff
@@ -244,13 +244,15 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
 		}, nil
 	}
 
-	instanceMeta, err := instanceMetadataGetter(node.Spec.ProviderID, node)
-	if err != nil {
-		utilruntime.HandleError(err)
-		return
+	if instanceMetadata == nil {
+		var err error
+		if instanceMetadata, err = instanceMetadataGetter(node.Spec.ProviderID, node); err != nil {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:249]: remove `instanceMetadataGetter` and call getInstanceMetadata directly?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 5]
```diff
@@ -244,13 +244,15 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
 		}, nil
 	}
 
-	instanceMeta, err := instanceMetadataGetter(node.Spec.ProviderID, node)
-	if err != nil {
-		utilruntime.HandleError(err)
-		return
+	if instanceMetadata == nil {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:247]: Thinking that the nil case should always result in an error and we should ensure that `getInstanceMetadata` is called prior to calling `updateNodeAddress`.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 4 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:249]: There is a difference between them: instanceMetadataGetter only get instance nodeAddresses, but getInstanceMetadata get both instance type and nodeAddresses.

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:249]: Maybe an option whether to get instance type make sense.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 3 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:424]: See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L415-L436
We should handle providerID not set scenario and get this modifier before get instance metadata that's why I move this out and also pass providerID and related modifier.

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:400]: The modifer is very simple though:
```
modifier := func(n *v1.Node) {
    if n.Spec.ProviderID == "" {
        n.Spec.ProviderID = providerID
    }
}
````

We can define the modifer here if providerID is "".

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 5 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: Updated, maybe better now.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 3 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: Thanks, looks good, simplified your suggestion and updated.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 6]
```diff
@@ -405,63 +393,16 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
...
-
+func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(
+	ctx context.Context,
+	node *v1.Node,
+	providerID string,
+	instanceMeta *cloudprovider.InstanceMetadata,
+) ([]nodeModifier, error) {
+
+	var nodeModifiers []nodeModifier
 	if node.Spec.ProviderID == "" {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:404]: There also needs to be a check for when the passed in providerID is "" which can happen if the provider returns `cloudprovider.NotImplemented` like in line 423

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 7]
```diff
@@ -513,6 +454,55 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
...
+			// do not, the taint will be removed, and this will not be retried
+			return "", err
+		}
+	}
+
+	return node.Spec.ProviderID, nil
+}
+
+// getInstanceMetadata get instance type and nodeAddresses, use Instances if InstancesV2 is off.
+func (cnc *CloudNodeController) getInstanceMetadata(ctx context.Context, providerID string, nodeName string, needType bool) (*cloudprovider.InstanceMetadata, error) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:479]: I'm not sure about having `needType` here because when v2 is enabled the type is always returned which can be confusing.

**Comment by andrewsykim** [comment]
FYI https://github.com/kubernetes/kubernetes/pull/93569

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 6 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:404]: `if node.Spec.ProviderID == "" && providerID != ""`, right?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 7 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:479]: How about always getting instance type?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 6 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:404]: Done.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 7 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: > How about always getting instance type?

Done.

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: yeah or maybe a `getNodeAddress` method that only sets `NodeAddress` in `InstanceMetadata`.

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: Not sure yet actually -- let's leave it as is for now (always return instance type in getInstanceMetadata)

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 8]
```diff
@@ -515,6 +457,53 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
...
+	// If InstancesV2 not implement, use Instances.
+	instances, ok := cnc.cloud.Instances()
+	if !ok {
+		return nil, fmt.Errorf("failed to get instances from cloud provider")
+	}
+	nodeAddresses, err := getNodeAddressesByProviderIDOrName(ctx, instances, node.Spec.ProviderID, node.Name)
+	if err != nil {
+		return nil, err
+	}
+	instanceType, err := getInstanceTypeByProviderIDOrName(ctx, instances, node.Spec.ProviderID, node.Name)
```
[the remark this answers is not in the document]
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:496]: Agree with you as we can't drop legacy Instances in a short time. Compared with reducing a http request in a frequently called function, it's worth to duplicate it.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 9]
```diff
@@ -515,6 +457,53 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
 	return nodeModifiers, nil
 }
 
+func (cnc *CloudNodeController) getProviderID(ctx context.Context, node *v1.Node) (string, error) {
+	if node.Spec.ProviderID == "" {
```
[the remark this answers is not in the document]
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: Done, also simplify following if and else conditions to `return providerID, err`, not sure comments in else condition should be adjusted.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 10]
```diff
@@ -515,6 +457,74 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
...
+	}
+
+	// if the cloud provider being used supports provider IDs, we want
+	// to propagate the error so that we re-try in the future; if we
+	// do not, the taint will be removed, and this will not be retried
+	return providerID, err
+}
+
+// getInstanceMetadata get instance type and nodeAddresses, use Instances if InstancesV2 is off.
+func (cnc *CloudNodeController) getInstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
```
Comment (gongguan) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:481]: @andrewsykim  one concern here: when providerID not passed here, `InstanceMetadata`, `getNodeAddressesByProviderIDOrName` and `getInstanceTypeByProviderIDOrName` could still work fine, right?

**Comment by gongguan** [comment]
Ping @andrewsykim , could you please take another look?

**Comment by andrewsykim** [comment]
Sorry for the delay @gongguan -- I'll review this soon

**Comment by gongguan** [comment]
> Sorry for the delay @gongguan -- I'll review this soon

No sorry, there is something wrong I'll fix it now, you can review after ci passed.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 10 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:482]: > @andrewsykim one concern here

No more since I set ProviderID in line 474.

**Comment by gongguan** [comment]
@andrewsykim ready for review now :)

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 11]
```diff
@@ -155,7 +155,12 @@ func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) {
 	}
 
 	for i := range nodes.Items {
-		cnc.updateNodeAddress(ctx, &nodes.Items[i])
+		instanceMetadata, err := cnc.getInstanceNodeAddresses(ctx, &nodes.Items[i])
+		if err != nil {
+			klog.Errorf("Error get instance metadata: %v", err)
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:160]: nit: "Error getting instance metadata for node addresses: %v"

**Review by andrewsykim — COMMENTED** [review]
LGTM, minor comment about error logging

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 11 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: done

**Comment by gongguan** [comment]
> LGTM, minor comment about error logging

@andrewsykim addressed and rebased, PTAL.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 12]
```diff
@@ -407,86 +395,16 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
...
+) ([]nodeModifier, error) {
 
-		return instanceMetadata, nil
-	}
-
-	instanceMeta, err := instanceMetadataGetter(providerID, node.Name, node)
-	if err != nil {
-		return nil, err
+	var nodeModifiers []nodeModifier
+	if node.Spec.ProviderID != providerID {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:406]: I think this check needs to be updated to
```
if node.Spec.ProviderID == "" && providerID != "" {
```

otherwise we might try to set an existing `node.Spec.ProviderID` to "". What do you think?

**Review by andrewsykim — COMMENTED** [review]
Sorry, had one more comment

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 12 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: Done, your suggestion is more legible.

> otherwise we might try to set an existing `node.Spec.ProviderID` to "". What do you think?

No, in line 454(getProviderID function), providerID will be set as node.Spec.ProviderID if node.Spec.ProviderID not nil. That means node.Spec.ProviderID not equal to providerID only occurs when node.Spec.ProviderID not set.

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: > No, in line 454(getProviderID function), providerID will be set as node.Spec.ProviderID if node.Spec.ProviderID not nil. That means node.Spec.ProviderID not equal to providerID only occurs when node.Spec.ProviderID not set.

Good point

**Comment by gongguan** [comment]
`should set TCP CLOSE_WAIT timeout [Privileged]` failed.

/test pull-kubernetes-e2e-kind-ipv6

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 13]
```diff
@@ -155,7 +155,12 @@ func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) {
 	}
 
 	for i := range nodes.Items {
-		cnc.updateNodeAddress(ctx, &nodes.Items[i])
+		instanceMetadata, err := cnc.getInstanceNodeAddresses(ctx, &nodes.Items[i])
+		if err != nil {
+			klog.Errorf("Error getting instance metadata for node addresses: %v", err)
+			return
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:161]: this should probably be `continue`?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 14]
```diff
@@ -363,9 +341,19 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
 		return
 	}
 
-	// TODO: getNodeModifiersFromCloudProvider and updateNodeAddress both call cloud api to get instanceMetadata,
-	// get instanceMetadata and pass it to getNodeModifiersFromCloudProvider and updateNodeAddress which reduces api calls.
-	nodeModifiers, err := cnc.getNodeModifiersFromCloudProvider(ctx, curNode)
+	providerID, err := cnc.getProviderID(ctx, curNode)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to initialize node %s at cloudprovider: %v", node.Name, err))
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:346]: should error message should probably be specific to getting the providerID:
```
utilruntime.HandleError(fmt.Errorf("failed to get provider ID for node %s: %v", node.Name, err))
```

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 15]
```diff
@@ -363,9 +341,19 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
...
+		return
+	}
+
+	instanceMetadata, err := cnc.getInstanceMetadata(ctx, providerID, curNode)
+	if err != nil {
+		utilruntime.HandleError(err)
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:352]: Wrap this error similar to other calls to HandleError here
```
utilruntime.HandleError(fmt.Errorf("failed to get instance metadata for node %s: %v", node.Name, err))
```

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 13 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: You're right, fixed.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 14 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: done

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 15 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: done

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

Thanks @gongguan
